### PR TITLE
Guard functions

### PR DESCRIPTION
**Pull request by lastmjs**
Closes https://github.com/demergent-labs/azle/issues/1209

### CODE REVIEW

File: examples/guard_functions/src/guards.ts [hunk 1]
```diff
@@ -54,38 +35,19 @@ class CustomError extends Error {
...
-    console.log('returnNonGuardResultObject called');
-    // @ts-ignore
-    return { badProp: 'Something other than a guard result' };
-}
-
-export function returnNonNullOkValue(): GuardResult {
-    console.log('nonNullOkValue called');
-    // @ts-ignore
-    return { Ok: 'Something other than null' };
+    return 'Upgrades to this canister are disabled';
```
Comment (dansteren) [inline examples/guard_functions/src/guards.ts:47]: Shouldn't this `throw`?

File: examples/guard_functions/src/invalid.ts
Comment (dansteren) [inline examples/guard_functions/src/invalid.ts:1]: It seems like many of these edge cases aren't valid with the new syntax. However, I wonder if there are some that still are, or if we have even different edge cases. I like that this file gets us thinking about all the ways users might do something wrong. For example, what if they did `@query([], bool, { guard: someInvalidIdentifier })`, what error would they get? I just want to make sure that users get direct, informative error messages when they do something wrong, so I would suggest that we modify this file for the scenarios that make sense with our new syntax, or that we create an issue for improving these errors later.

File: src/compiler/typescript_to_rust/azle_generate_rearchitecture/src/main.rs [hunk 2]
```diff
@@ -104,11 +105,20 @@ fn main() -> Result<(), String> {
...
+            let (guard_attribute, guard_function) = if let Some(guard_name) = &canister_method.guard_name
+            {
+                get_guard_token_stream(guard_name)
+            } else {
+                (quote!(), quote!())
+            };
+
             quote! {
-                #[ic_cdk_macros::query(manual_reply = true, composite = #is_composite)]
+                #[ic_cdk_macros::query(manual_reply = true, composite = #is_composite #guard_attribute)]
```
Comment (dansteren) [inline src/compiler/typescript_to_rust/azle_generate_rearchitecture/src/main.rs:116]: Is there supposed to be comma between composite and #guard_attribute? I'm sure you did it right but it caught my eye so just double check.

**Review by dansteren — APPROVED** [review]
Other than the comments I left it looks good.

File: examples/guard_functions/src/guards.ts (hunk 1 above)
Comment (lastmjs) [reply examples/guard_functions/src/guards.ts:47]: Not needed I'll just delete it in my next PR

File: examples/guard_functions/src/invalid.ts
Comment (lastmjs) [reply examples/guard_functions/src/invalid.ts:1]: I'm just not sure this makes sense, an invalid identifier would show up as a typescript or javascript error, we don't do many of the things we used to do so I would say it's more likely not our concern.

Generally we want errors to show nicely but it's so different now.

File: src/compiler/typescript_to_rust/azle_generate_rearchitecture/src/main.rs (hunk 2 above)
Comment (lastmjs) [reply src/compiler/typescript_to_rust/azle_generate_rearchitecture/src/main.rs:116]: It's correct, the comma is in the guard_attribute

**Pull request merged**